### ISSUE TITLE
OCRmyPDF 10.1.0 sets OMP_THREAD_LIMIT to an invalid value

### ISSUE THREAD

**Issue by rbrito** [jbarlow83/OCRmyPDF#580]
Hi, @jbarlow83.

I started getting messages like this since version 10:

```
    1 [tesseract] libgomp: Invalid value for environment variable OMP_THREAD_LIMIT                                                                                           
```

I digged into the code and, in https://github.com/jbarlow83/OCRmyPDF/blob/24b6a4ad502875986fa85ec184b71d511d783202/src/ocrmypdf/builtin_plugins/tesseract_ocr.py#L119-L135 makes a calculation that I believe is wrong.

I added a bunch of debugging statements and I got this:

```
     **************** value of OMP_THREAD_LIMIT: <>
     **************** options.jobs= 4
     **************** len(pdfinfo) = 619
     **************** options.jobs // len(pdfinfo) = 0
     **************** going to set OMP_THREAD_LIMIT: 0
```
Of course, computing `min(3, options.jobs // len(pdfinfo), len(pdfinfo))` is going to result in 0 for many, many cases (especially for those where people are OCR'ing books of many pages), since it is very likely that options.jobs/len(pdfinfo) < 1, whose floor is going to be 0.

Perhaps you meant to write `tess_threads = max(3, options.jobs // len(pdfinfo))` instead?

##### Sanity check

```python
ncoress = [1, 2, 4, 6, 8, 16]
npagess = [1, 2, 3, 15, 100]

def tess_threads(cores, npages):
     return max(3, cores // npages)

for ncores in ncoress:
     for npages in npagess:
         threads = tess_threads(ncores, npages)
         print(f'For {ncores} cores and {npages} pages, use {threads} threads for tesseract (total threads spawned = {threads * npages}).')
```

which results in:

```
For 1 cores and 1 pages, use 3 threads for tesseract (total threads spawned = 3).
For 1 cores and 2 pages, use 3 threads for tesseract (total threads spawned = 6).
For 1 cores and 3 pages, use 3 threads for tesseract (total threads spawned = 9).
For 1 cores and 15 pages, use 3 threads for tesseract (total threads spawned = 45).
For 1 cores and 100 pages, use 3 threads for tesseract (total threads spawned = 300).
For 2 cores and 1 pages, use 3 threads for tesseract (total threads spawned = 3).
For 2 cores and 2 pages, use 3 threads for tesseract (total threads spawned = 6).
For 2 cores and 3 pages, use 3 threads for tesseract (total threads spawned = 9).
For 2 cores and 15 pages, use 3 threads for tesseract (total threads spawned = 45).
For 2 cores and 100 pages, use 3 threads for tesseract (total threads spawned = 300).
For 4 cores and 1 pages, use 4 threads for tesseract (total threads spawned = 4).
For 4 cores and 2 pages, use 3 threads for tesseract (total threads spawned = 6).
For 4 cores and 3 pages, use 3 threads for tesseract (total threads spawned = 9).
For 4 cores and 15 pages, use 3 threads for tesseract (total threads spawned = 45).
For 4 cores and 100 pages, use 3 threads for tesseract (total threads spawned = 300).
For 6 cores and 1 pages, use 6 threads for tesseract (total threads spawned = 6).
For 6 cores and 2 pages, use 3 threads for tesseract (total threads spawned = 6).
For 6 cores and 3 pages, use 3 threads for tesseract (total threads spawned = 9).
For 6 cores and 15 pages, use 3 threads for tesseract (total threads spawned = 45).
For 6 cores and 100 pages, use 3 threads for tesseract (total threads spawned = 300).
For 8 cores and 1 pages, use 8 threads for tesseract (total threads spawned = 8).
For 8 cores and 2 pages, use 4 threads for tesseract (total threads spawned = 8).
For 8 cores and 3 pages, use 3 threads for tesseract (total threads spawned = 9).
For 8 cores and 15 pages, use 3 threads for tesseract (total threads spawned = 45).
For 8 cores and 100 pages, use 3 threads for tesseract (total threads spawned = 300).
For 16 cores and 1 pages, use 16 threads for tesseract (total threads spawned = 16).
For 16 cores and 2 pages, use 8 threads for tesseract (total threads spawned = 16).
For 16 cores and 3 pages, use 5 threads for tesseract (total threads spawned = 15).
For 16 cores and 15 pages, use 3 threads for tesseract (total threads spawned = 45).
For 16 cores and 100 pages, use 3 threads for tesseract (total threads spawned = 300).
```

which looks sensible to me...

What do you think? Did I miss something completely obvious? I hope that I am not going to put myself into public shame...

Rogério Brito

**Comment by jbarlow83** [comment]
We need 1 <= OMP_THREAD_LIMIT <= 3. The 0 is wrong, true.

The key finding from performance testing is that Tesseract's multithreading doesn't help that much. It's a historically singled threaded program with multithreading bolted in a few key places. We get better results from parallelizing a single threaded Tesseract process.

The only time Tesseract threads do help is when we have more CPU cores available than pages (since we parallelize over pages). Then we can turn it up above one. But we don't want to overrun the CPU. And also from testing, turning it up above 3 doesn't help (wastes energy, takes just as long).

The decision about how many page workers gets is made in `_sync.py::run_pipeline`.

**Comment by jbarlow83** [comment]
Fixed in v10.1.1. Too many zeros and ones.

**Comment by jonaswinkler** [comment]
I'd like to mention that the documentation still states that OMP_THREAD_LIMIT is always set to 1.

https://ocrmypdf.readthedocs.io/en/latest/advanced.html#overriding-default-tesseract
